rmf_task_ros2/src/rmf_task_ros2/dispatcher/Dispatcher.cpp

```diff
@@ -0,0 +1,323 @@
+    action_client->on_terminate(
+      std::bind(&Implementation::terminate_task, this, _1));
+    action_client->on_change(
+      std::bind(&Implementation::task_status_cb, this, _1));
+  }
+
+  TaskID submit_task(const TaskProfile& task)
+  {
+    auto submitted_task = task;
+
+    // auto generate a task_id for a given submitted task
+    // TODO: generate a unique task_id based of clock
+    submitted_task.task_id =
+      std::to_string( ((int)task.task_type.type)*1000 + (i++) );
+    // "task" + std::to_string((int)task.task_type)
+    //   + "-" + std::to_string((int)(node->now().seconds()));
+
+    submitted_task.submission_time = node->now();
+
+    // add task to internal cache
+    TaskStatus status;
+    status.task_profile = submitted_task;
+    auto new_task_status = std::make_shared<TaskStatus>(status);
+    (*active_dispatch_tasks)[submitted_task.task_id] = new_task_status;
+
+    if (on_change_fn)
+      on_change_fn(new_task_status);
+
+    bidding::BidNotice bid_notice;
+    bid_notice.task_profile = submitted_task;
+    bid_notice.time_window = rmf_traffic_ros2::convert(
+      rmf_traffic::time::from_seconds(bidding_time_window));
+    auctioneer->start_bidding(bid_notice);
+
+    return submitted_task.task_id;
+  }
+
+  bool cancel_task(const TaskID& task_id)
+  {
+    // check if key exists
+    if (!(*active_dispatch_tasks).count(task_id))
+      return false;
+
+    TaskProfile profile;
+    {
+      // make sure status will expired, todo: cleaner way
+      auto cancel_task_status = (*active_dispatch_tasks)[task_id];
+      profile = cancel_task_status->task_profile;
+
+      // Cancel bidding. This will remove the bidding process
+      if (cancel_task_status->state == TaskStatus::State::Pending)
+      {
+        cancel_task_status->state = TaskStatus::State::Canceled;
+        terminate_task(cancel_task_status);
+
+        if (on_change_fn)
+          on_change_fn(cancel_task_status);
+
+        return true;
+      }
+    }
+
+    // Cancel action task, this will only send a cancel to FA. up to
+    // the FA whether to cancel the task. On change is implemented
+    // internally in action client
+    return action_client->cancel_task(profile);
+  }
+
+  rmf_utils::optional<TaskStatus::State> get_task_state(
+    const TaskID& task_id)
+  {
+    // check if key doesnt exist
+    if ((*active_dispatch_tasks).count(task_id))
+      return (*active_dispatch_tasks)[task_id]->state;
+
+    if ((*terminal_dispatch_tasks).count(task_id))
+      return (*terminal_dispatch_tasks)[task_id]->state;
+
+    return rmf_utils::nullopt;
+  }
+
+  void receive_bidding_winner_cb(
+    const TaskID& task_id,
+    const rmf_utils::optional<bidding::Submission> winner)
+  {
+    if (!(*active_dispatch_tasks).count(task_id))
+      return;
+
+    std::cout << "[Dispatch::BiddingResultCb] | task: " << task_id;
+    auto pending_task_status = (*active_dispatch_tasks)[task_id];
```

Comment:
Nitpick: We can eliminate a redundant map lookup between this line and 147 by doing

```
const auto it = active_dispatch_tasks->find(task_id);
if (it == active_dispatch_tasks->end())
  return;

const auto& pending task_status = it->second;
```